[interpreter] Change how wast.js is built by using Js_of_ocaml

takikawa opened the pull request:
See issue #1506 for a discussion on the wast.js build mode state.

This PR replaces the current JS build mode for wast.js by using dune and [Js_of_ocaml](https://ocsigen.org/js_of_ocaml/latest/manual/overview) to compile the reference interpreter to JS.

takikawa commented:
Looks like this overlaps a bit with #1408 but is more minimally targeted at just the JS support. This PR also does the ArrayBuffer conversion for the JS interface, which wasn't in the other PR AFAIK.

rossberg commented on interpreter/Makefile:
```diff
@@ -109,12 +109,9 @@ _build/$(LIB).cmxa: $(FILES) $(LIB).mllib _tags Makefile
-		(cd _build/jslib; ./build.sh ../../$@)
+$(JSLIB):
+		dune build wast.bc.js
+		cp ../_build/default/interpreter/wast.bc.js ./wast.js
```
Can you introduce proper build rules for wast.bc.js, so that dependencies and cleanup work?

takikawa replied:
This should be addressed now. I redid the Makefile rule to avoid using dune and just use `ocamlbuild` and calling `js_of_ocaml` directly (there is an ocamlbuild plugin for `js_of_ocaml` but I was unable to get it to work correctly). It should do cleanup now for example.

Happy to fine tune the build rules if anything looks off still.

rossberg commented on interpreter/meta/jslib/wast.ml:
```diff
@@ -0,0 +1,25 @@
+        String.iteri (fun i c -> Typed_array.set u8arr i (int_of_char c)) bs; buf
+
+      method decode buf = decode buf 80
+      method decode2 buf width = decode buf width
```
Can we drop this duplication and always require the width? I don't think the minor convenience justifies it, and 80 is a fairly random default these days.

takikawa replied:
Done.

rossberg commented on interpreter/meta/jslib/wast.ml:
```diff
@@ -0,0 +1,25 @@
+open Js_of_ocaml
```
If you introduce this as wast.ml, shouldn't you remove the prior wasm.ml?

takikawa replied:
I've now removed `wasm.ml` and associated build files in the latest commits.

rossberg reviewed (commented):
Thanks for looking into this! I'm on board with replacing BuckleScript.

takikawa commented:
I think the latest push should address the review comments. In addition, I modified the README to match the new setup and added CI integration that will build `wast.js` along with the normal interpreter build.

rossberg commented on interpreter/README.md:
````diff
@@ -160,6 +163,20 @@ WebAssemblyText.decode(binary)
+
+```
+let wast = require("./wast.js");
+let binary = wast.WebAssemblyText.encode("(module"));
````
```suggestion
let binary = wast.WebAssemblyText.encode("(module)");
```

rossberg commented on interpreter/README.md:
````diff
@@ -160,6 +163,20 @@ WebAssemblyText.decode(binary)
+
+```
+load("./wast.js");
+let binary = WebAssemblyText.encode("(module"));
````
```suggestion
let binary = WebAssemblyText.encode("(module)");
```

rossberg commented on interpreter/meta/jslib/wast.ml:
```diff
@@ -0,0 +1,26 @@
+        let u8arr = new%js Typed_array.uint8Array_fromBuffer buf in
+        String.iteri (fun i c -> Typed_array.set u8arr i (int_of_char c)) bs; buf
+
+      method decode (buf : Typed_array.arrayBuffer Js.t) width : (Js.js_string Js.t) =
```
Perhaps by declaring the parameter
```
(optwidth : int Js.optdef)
```
and then doing
```
let width = Js.Optdef.get optwidth (fun () -> 80) in
```
it is possible to emulate the previous default parameter?

hhugo replied:
You can achieve that behavior using JS.Unsafe.* functions.
Maybe with `val decode = Js.Unsafe.callback (fun bud width -> ...)`

rossberg reviewed (commented):
This is great!

takikawa commented on interpreter/Makefile:
```diff
@@ -108,14 +109,12 @@ _build/$(LIB).cmxa: $(FILES) $(LIB).mllib _tags Makefile
-		(cd _build/jslib; ./build.sh ../../$@)
+.INTERMEDIATE:	$(JSLIB).byte
+$(JSLIB).byte:	meta/jslib/$(JSLIB).ml
+		$(OCBA) -use-ocamlfind -pkg js_of_ocaml -pkg js_of_ocaml-ppx meta/jslib/$(JSLIB).byte
```
I was able to refactor this a bit so that it uses `$@` as the target now.

rossberg reviewed (approved):
Sorry, missed your last reply.

LGTM, thanks!
